{"identifier":{"interfaceLanguage":"swift","url":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/Validation"},"seeAlsoSections":[{"anchor":"Validation-and-Errors","identifiers":["doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ValidationError","doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/CleanExit","doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ExitCode"],"title":"Validation and Errors","generated":true}],"primaryContentSections":[{"kind":"content","content":[{"type":"heading","text":"Overview","level":2,"anchor":"Overview"},{"inlineContent":[{"text":"While ","type":"text"},{"code":"ArgumentParser","type":"codeVoice"},{"text":" validates that the inputs given by your user match the requirements and types that you define in each command, there are some requirements that can’t easily be described in Swift’s type system, such as the number of elements in an array, or an expected integer value.","type":"text"}],"type":"paragraph"},{"anchor":"Validating-Command-Line-Input","level":3,"type":"heading","text":"Validating Command-Line Input"},{"type":"paragraph","inlineContent":[{"type":"text","text":"To validate your commands properties after parsing, implement the "},{"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ParsableArguments\/validate()-5r0ge","isActive":true,"type":"reference"},{"text":" method on any ","type":"text"},{"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ParsableCommand","isActive":true,"type":"reference"},{"text":", ","type":"text"},{"type":"reference","identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ParsableArguments","isActive":true},{"text":", or ","type":"text"},{"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/AsyncParsableCommand","type":"reference","isActive":true},{"text":" type. Throwing an error from the ","type":"text"},{"code":"validate()","type":"codeVoice"},{"text":" method causes the program to print a message to standard error and exit with an error code, preventing the ","type":"text"},{"code":"run()","type":"codeVoice"},{"type":"text","text":" method from being called with invalid inputs."}]},{"type":"paragraph","inlineContent":[{"type":"text","text":"Here’s a command that prints out one or more random elements from the list you provide. Its "},{"type":"codeVoice","code":"validate()"},{"type":"text","text":" method catches three different errors that a user can make and throws a relevant error for each one."}]},{"code":["struct Select: ParsableCommand {","    @Option var count: Int = 1","    @Argument var elements: [String] = []","","    mutating func validate() throws {","        guard count >= 1 else {","            throw ValidationError(\"Please specify a 'count' of at least 1.\")","        }","","        guard !elements.isEmpty else {","            throw ValidationError(\"Please provide at least one element to choose from.\")","        }","","        guard count <= elements.count else {","            throw ValidationError(\"Please specify a 'count' less than the number of elements.\")","        }","    }","","    mutating func run() {","        print(elements.shuffled().prefix(count).joined(separator: \"\\n\"))","    }","}"],"syntax":"swift","type":"codeListing"},{"inlineContent":[{"type":"text","text":"When you provide useful error messages, they can guide new users to success with your command-line tool!"}],"type":"paragraph"},{"type":"codeListing","code":["% select","Error: Please provide at least one element to choose from.","Usage: select [--count <count>] [<elements> ...]","  See 'select --help' for more information.","% select --count 2 hello","Error: Please specify a 'count' less than the number of elements.","Usage: select [--count <count>] [<elements> ...]","  See 'select --help' for more information.","% select --count 0 hello hey hi howdy","Error: Please specify a 'count' of at least 1.","Usage: select [--count <count>] [<elements> ...]","  See 'select --help' for more information.","% select --count 2 hello hey hi howdy","howdy","hey"],"syntax":null},{"anchor":"Handling-Post-Validation-Errors","type":"heading","level":2,"text":"Handling Post-Validation Errors"},{"type":"paragraph","inlineContent":[{"type":"text","text":"The "},{"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ValidationError","isActive":true,"type":"reference"},{"text":" type is a special ","type":"text"},{"type":"codeVoice","code":"ArgumentParser"},{"text":" error — a validation error’s message is always accompanied by an appropriate usage string. You can throw other errors, from either the ","type":"text"},{"type":"codeVoice","code":"validate()"},{"type":"text","text":" or "},{"type":"codeVoice","code":"run()"},{"text":" method to indicate that something has gone wrong that isn’t validation-specific. Errors that conform to ","type":"text"},{"type":"codeVoice","code":"CustomStringConvertible"},{"text":" or ","type":"text"},{"code":"LocalizedError","type":"codeVoice"},{"type":"text","text":" provide the best experience for users."}]},{"syntax":"swift","type":"codeListing","code":["struct LineCount: ParsableCommand {","    @Argument var file: String","","    mutating func run() throws {","        let contents = try String(contentsOfFile: file, encoding: .utf8)","        let lines = contents.split(separator: \"\\n\")","        print(lines.count)","    }","}"]},{"inlineContent":[{"text":"The throwing ","type":"text"},{"code":"String(contentsOfFile:encoding:)","type":"codeVoice"},{"type":"text","text":" initializer fails when the user specifies an invalid file. "},{"type":"codeVoice","code":"ArgumentParser"},{"text":" prints its error message to standard error and exits with an error code.","type":"text"}],"type":"paragraph"},{"type":"codeListing","syntax":null,"code":["% line-count file1.swift","37","% line-count non-existing-file.swift","Error: The file “non-existing-file.swift” couldn’t be opened because","there is no such file."]},{"type":"paragraph","inlineContent":[{"text":"If you print your error output yourself, you still need to throw an error from ","type":"text"},{"type":"codeVoice","code":"validate()"},{"type":"text","text":" or "},{"type":"codeVoice","code":"run()"},{"type":"text","text":", so that your command exits with the appropriate exit code. To avoid printing an extra error message, use the "},{"type":"codeVoice","code":"ExitCode"},{"type":"text","text":" error, which has static properties for success, failure, and validation errors, or lets you specify a specific exit code."}]},{"type":"codeListing","syntax":"swift","code":["struct RuntimeError: Error, CustomStringConvertible {","    var description: String","}","","struct Example: ParsableCommand {","    @Argument var inputFile: String","","    mutating func run() throws {","        if !ExampleCore.processFile(inputFile) {","            \/\/ ExampleCore.processFile(_:) prints its own errors","            \/\/ and returns `false` on failure.","            throw ExitCode.failure","        }","    }","}"]},{"level":2,"text":"Handling Transform Errors","type":"heading","anchor":"Handling-Transform-Errors"},{"type":"paragraph","inlineContent":[{"text":"During argument and option parsing, you can use a closure to transform the command line strings to custom types. If this transformation fails, you can throw a ","type":"text"},{"type":"codeVoice","code":"ValidationError"},{"type":"text","text":"; its "},{"type":"codeVoice","code":"message"},{"type":"text","text":" property will be displayed to the user."}]},{"type":"paragraph","inlineContent":[{"text":"In addition, you can throw your own errors. Errors that conform to ","type":"text"},{"type":"codeVoice","code":"CustomStringConvertible"},{"text":" or ","type":"text"},{"code":"LocalizedError","type":"codeVoice"},{"type":"text","text":" provide the best experience for users."}]},{"code":["struct ExampleTransformError: Error, CustomStringConvertible {","  var description: String","}","","struct ExampleDataModel: Codable {","  let identifier: UUID","  let tokens: [String]","  let tokenCount: Int","","  static func dataModel(_ jsonString: String) throws -> ExampleDataModel  {","    guard let data = jsonString.data(using: .utf8) else { throw ValidationError(\"Badly encoded string, should be UTF-8\") }","    return try JSONDecoder().decode(ExampleDataModel.self, from: data)","  }","}","","struct Example: ParsableCommand {","  \/\/ Reads in the argument string and attempts to transform it to","  \/\/ an `ExampleDataModel` object using the `JSONDecoder`. If the","  \/\/ string is not valid JSON, `decode` will throw an error and","  \/\/ parsing will halt.","  @Argument(transform: ExampleDataModel.dataModel)","  var inputJSON: ExampleDataModel","","  \/\/ Specifiying this option will always cause the parser to exit","  \/\/ and print the custom error.","  @Option(transform: { throw ExampleTransformError(description: \"Trying to write to failOption always produces an error. Input: \\($0)\") })","  var failOption: String?","}"],"type":"codeListing","syntax":"swift"},{"type":"paragraph","inlineContent":[{"type":"text","text":"Throwing from a transform closure benefits users by providing context and can reduce development time by pinpointing issues quickly and more precisely."}]},{"code":["% example '{\"Bad JSON\"}'","Error: The value '{\"Bad JSON\"}' is invalid for '<input-json>': dataCorrupted(Swift.DecodingError.Context(codingPath: [], debugDescription: \"The given data was not valid JSON.\", underlyingError: Optional(Error Domain=NSCocoaErrorDomain Code=3840 \"No value for key in object around character 11.\" UserInfo={NSDebugDescription=No value for key in object around character 11.})))","Usage: example <input-json> --fail-option <fail-option>","  See 'select --help' for more information."],"syntax":null,"type":"codeListing"},{"type":"paragraph","inlineContent":[{"type":"text","text":"While throwing standard library or Foundation errors adds context, custom errors provide the best experience for users and developers."}]},{"syntax":null,"code":["% example '{\"tokenCount\":0,\"tokens\":[],\"identifier\":\"F77D661C-C5B7-448E-9344-267B284F66AD\"}' --fail-option=\"Some Text Here!\"","Error: The value 'Some Text Here!' is invalid for '--fail-option <fail-option>': Trying to write to failOption always produces an error. Input: Some Text Here!","Usage: example <input-json> --fail-option <fail-option>","  See 'select --help' for more information."],"type":"codeListing"}]}],"kind":"article","variants":[{"traits":[{"interfaceLanguage":"swift"}],"paths":["\/documentation\/argumentparser\/validation"]}],"sections":[],"metadata":{"title":"Providing Custom Validation","modules":[{"name":"ArgumentParser"}],"roleHeading":"Article","role":"article"},"schemaVersion":{"major":0,"minor":3,"patch":0},"hierarchy":{"paths":[["doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser"]]},"abstract":[{"type":"text","text":"Provide helpful feedback to users when things go wrong."}],"references":{"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/ExitCode":{"abstract":[{"text":"An error type that only includes an exit code.","type":"text"}],"type":"topic","identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ExitCode","title":"ExitCode","role":"symbol","url":"\/documentation\/argumentparser\/exitcode","navigatorTitle":[{"kind":"identifier","text":"ExitCode"}],"fragments":[{"kind":"keyword","text":"struct"},{"kind":"text","text":" "},{"text":"ExitCode","kind":"identifier"}],"kind":"symbol"},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/ValidationError":{"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ValidationError","type":"topic","navigatorTitle":[{"kind":"identifier","text":"ValidationError"}],"fragments":[{"text":"struct","kind":"keyword"},{"text":" ","kind":"text"},{"text":"ValidationError","kind":"identifier"}],"url":"\/documentation\/argumentparser\/validationerror","kind":"symbol","role":"symbol","abstract":[{"type":"text","text":"An error type that is presented to the user as an error with parsing their"},{"text":" ","type":"text"},{"type":"text","text":"command-line input."}],"title":"ValidationError"},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser":{"title":"ArgumentParser","type":"topic","abstract":[{"text":"Straightforward, type-safe argument parsing for Swift.","type":"text"}],"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser","role":"collection","url":"\/documentation\/argumentparser","kind":"symbol"},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/ParsableArguments":{"url":"\/documentation\/argumentparser\/parsablearguments","navigatorTitle":[{"kind":"identifier","text":"ParsableArguments"}],"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ParsableArguments","role":"symbol","title":"ParsableArguments","kind":"symbol","fragments":[{"text":"protocol","kind":"keyword"},{"text":" ","kind":"text"},{"text":"ParsableArguments","kind":"identifier"}],"type":"topic","abstract":[{"type":"text","text":"A type that can be parsed from a program’s command-line arguments."}]},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/CleanExit":{"navigatorTitle":[{"text":"CleanExit","kind":"identifier"}],"abstract":[{"type":"text","text":"An error type that represents a clean (i.e. non-error state) exit of the"},{"text":" ","type":"text"},{"type":"text","text":"utility."}],"type":"topic","title":"CleanExit","identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/CleanExit","url":"\/documentation\/argumentparser\/cleanexit","kind":"symbol","role":"symbol","fragments":[{"kind":"keyword","text":"struct"},{"text":" ","kind":"text"},{"text":"CleanExit","kind":"identifier"}]},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/ParsableArguments/validate()-5r0ge":{"required":true,"kind":"symbol","url":"\/documentation\/argumentparser\/parsablearguments\/validate()-5r0ge","fragments":[{"kind":"keyword","text":"func"},{"text":" ","kind":"text"},{"kind":"identifier","text":"validate"},{"text":"() ","kind":"text"},{"text":"throws","kind":"keyword"}],"abstract":[{"type":"text","text":"Validates the properties of the instance after parsing."}],"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ParsableArguments\/validate()-5r0ge","defaultImplementations":1,"type":"topic","title":"validate()","role":"symbol"},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/ParsableCommand":{"fragments":[{"kind":"keyword","text":"protocol"},{"kind":"text","text":" "},{"kind":"identifier","text":"ParsableCommand"}],"abstract":[{"text":"A type that can be executed as part of a nested tree of commands.","type":"text"}],"kind":"symbol","type":"topic","url":"\/documentation\/argumentparser\/parsablecommand","identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/ParsableCommand","navigatorTitle":[{"text":"ParsableCommand","kind":"identifier"}],"role":"symbol","title":"ParsableCommand"},"doc://swift-argument-parser.ArgumentParser/documentation/ArgumentParser/AsyncParsableCommand":{"identifier":"doc:\/\/swift-argument-parser.ArgumentParser\/documentation\/ArgumentParser\/AsyncParsableCommand","url":"\/documentation\/argumentparser\/asyncparsablecommand","type":"topic","abstract":[{"type":"text","text":"A type that can be executed asynchronously, as part of a nested tree of"},{"text":" ","type":"text"},{"text":"commands.","type":"text"}],"fragments":[{"kind":"keyword","text":"protocol"},{"text":" ","kind":"text"},{"text":"AsyncParsableCommand","kind":"identifier"}],"navigatorTitle":[{"text":"AsyncParsableCommand","kind":"identifier"}],"role":"symbol","kind":"symbol","title":"AsyncParsableCommand"}}}